hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,262 @@
 
 import asyncio
 import concurrent.futures._base as base
+from contextlib import nullcontext
 import multiprocessing as mp
 import multiprocessing.connection
-import time
-from functools import lru_cache
-from typing import Optional, Tuple, Generic, TypeVar
+import os
+import threading
+import uuid
+from enum import Enum, auto
+from typing import Generic, TypeVar, Dict, Optional, Any, Callable
 
-from hivemind.utils.threading import run_in_background
+import torch    # used for py3.7-compatible shared memory
```

Comment:
Still would prefer explicit imports from torch

Reply:
the explicit version imports
`from torch import empty, uint8`

which somehow looks even more confusing :)